###Titanic Python

1/May/2019

*"first catch your hare"*

In our terms get the data and understand the objective. 


In [1]:
# simple statment to confirm that the notebook processing is active
# for Google colab - it often takes a few seconds to get things started 
# this simple step with its clear output provides a visual check
print ("Titanic - Python")

Titanic - Python


### Imports

One of the strengths of Python is the range of external libraries which can be used.

An important mathematical library - which implements multidimensional arrays and many methods for manipulating them is numpy - which by convention we import as 'np'.  

For data analysis there is an important library called pandas which implements a dataframe which can be used to store tabular data and implements many useful methods for manipulating these dataframes - by convention we import it as 'pd'.



In [0]:
import pandas as pd
import numpy as np

from sklearn import preprocessing

import matplotlib.pyplot as plt 

import seaborn as sns

Some of our libraries - particularly the presentation ones do require some initial set-up
which is as well to do here with the import if it is 'standard' setup.

In [0]:
# mathplot sont size
plt.rc("font", size=14)

# seaborn styles
sns.set(style="white") #white background style for seaborn plots
sns.set(style="whitegrid", color_codes=True)

Import data from Google drive : 

For the next section of code to work you have to have copied the relevant files to your
Google Drive.

And as far as I know it only works from Google Colab - which is where I created this notebook.

I have commented the code to prevent it running and generating an error if the Google Drive link is 
not in place -  it requires a Google Drive authorisation every session.

Clearly if you uncomment and use this code then you need to skip or comment out the next section.




In [0]:
# from google.colab import drive
# drive.mount('/content/gdrive')

# !ls /content/gdrive
# !ls -l /content/gdrive/My\ Drive
# !ls -l /content/gdrive/My\ Drive/Colab\ Notebooks
# !ls -l /content/gdrive/My\ Drive/Colab\ Notebooks/Titanic

# titanic_known = pd.read_csv('/content/gdrive/My Drive/Colab Notebooks/Titanic/titanic_known.csv')
# print (titanic_known.describe())
# titanic_unknown = pd.read_csv('/content/gdrive/My Drive/Colab Notebooks/Titanic/titanic_unknown.csv')
# print (titanic_unknown.describe())

Import data from a web source

Pandas dataframes have a read_cvs method for reading csv sources - there are other simlar methods for reading different format sources all of which can take a web URL as the location as well as a local file.

The below is a simple load directly from a web source - which is fine for small datasets but not recommended if the amount of data gets large.

Any webserver will work.

I use github to store my working code - you can access files in a github repository directly using the raw.githubusercontent.com/*username*/. . . address. 

Github also provides a 'special' repository name of *username*.github.io to allow you to create a repository which works as a static webserver (i.e. no server-side processing) - and this is perfectly serviceable for loading data into an object using the pandas read_csv method.

After reading file into a pandas dateframe we use len to get the number of rows and the we use the datframe method count to get the number of entries in each collum - and the collum names - this is pretty much a  *'have I got the data I expect?'*  verification.  And it gives us some basic insight into our data.

Whereas the whole purpose of using a machine to analyse data is to avoid much of the associated drudge work - it is hugely beneficial to have a good understanding of the data itself - particularly important is to make sure that the meaning of data is precise - for example, in this case we will have cause to ask what exactly does the Fare column contain.  And the asnwer to that question gives us an additional piece of data that we don't know that we have at the moment and that improves the performance of some of our models.

To this end we will spend a bit of time just looking at the data in various ways.

In [7]:
titanic_known = pd.read_csv("https://raw.githubusercontent.com/johnsl01/Titanic_Python/master/titanic_known.csv", sep=",")

# titanic_known = pd.read_csv("https://johnsl01.github.io/titanic/titanic_known.csv", sep=",")
print ("Known :")
print (str(len(titanic_known)) + " Rows")
print (titanic_known.count())

titanic_unknown = pd.read_csv("https://johnsl01.github.io/titanic/titanic_unknown.csv", sep=",")
print ("\nUnknown :") 
print (str(len(titanic_unknown)) + " Rows")
print (titanic_unknown.count())

Known :
891 Rows
PassengerId    891
Survived       891
Pclass         891
Name           891
Sex            891
Age            714
SibSp          891
Parch          891
Ticket         891
Fare           891
Cabin          204
Embarked       889
dtype: int64

Unknown :
418 Rows
PassengerId    418
Pclass         418
Name           418
Sex            418
Age            332
SibSp          418
Parch          418
Ticket         418
Fare           417
Cabin           91
Embarked       418
dtype: int64


#### head and tail

Our data is a bit unreal at the moment - we know the column names and the count of data in each - and we know that some ages and a lot of cabins are missing.

And we know that the 'known' dataframe has an extra column; 'Survived' which we expect to tell us what the fate of that passenger was - which is obviously missing from the 'unknown' dataframe.

One of the simplest things to do with data is to inspect it row by row and the starting point in this is usually to use the head and tail methods.  Head and tail derive from two unix commands which appeared very early in unix - and were used to show the first and last lines in a file.  By default they returned 5 lines and this default has persisted for almost 50 years as they are reincarnated in different systems.

So pandas dataframes have a head and a tail method which do the same for dataframe rows : 

In [8]:
print ("Known : -  first 5 and last 2:")
print (titanic_known.head())
print (titanic_known.tail(2))

print ("UnkKnown : -  first 5 and last 2:")
print (titanic_unknown.head())
print (titanic_unknown.tail(2))

Known : -  first 5 and last 2:
   PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0            1         0       3  ...   7.2500   NaN         S
1            2         1       1  ...  71.2833   C85         C
2            3         1       3  ...   7.9250   NaN         S
3            4         1       1  ...  53.1000  C123         S
4            5         0       3  ...   8.0500   NaN         S

[5 rows x 12 columns]
     PassengerId  Survived  Pclass  ...   Fare Cabin  Embarked
889          890         1       1  ...  30.00  C148         C
890          891         0       3  ...   7.75   NaN         Q

[2 rows x 12 columns]
UnkKnown : -  first 5 and last 2:
   PassengerId  Pclass  ... Cabin Embarked
0          892       3  ...   NaN        Q
1          893       3  ...   NaN        S
2          894       2  ...   NaN        Q
3          895       3  ...   NaN        S
4          896       3  ...   NaN        S

[5 rows x 11 columns]
     PassengerId  Pclass                 

OK - so this tells us someting and also raises some concerns.

The encoding of the survived column is 1 and 0  (1 for survived and 0 for died) - this is the most common numeric for a boolean value - 'Y' and 'N' and 'T' and 'F' being the most common character encodings - it is important to be sure that we get the mapping correct.  

The Pclass column is the passenger class - the ship was divided into three different classes - 1st, 2nd and 3rd.  1st class passengers had luxurious cabins, they had open access to the upper decks and to the luxurious 'state' rooms of the ship - they were generally wealthier and as well as some having brought personal staff with them they had a lot of the ship's crew dedicated to their care.  We should expect that this will be a factor in survival - access and crew deference being significant factors. 2nd class had individual cabins but less deck access and restricted access to some areas.  3rd class passengers generally were below decks - had dormitory type accomodation and were packed in pretty tight - they did not have much deck access and were unable to get into much of the 1st and 2nd class areas.

The name is the passenger's name and includes a title Mr. Mrs. etc as well as maiden names for married women.  This is text data so while we will want to make use of it we have to be careful in how we processes it - we will definitely want to pull out the title as that will help us in some cases with filling in the missing ages - and we will make some use of the name information to try and derive some information to help us with understanding what size of a party each passenger was likely in.

Sex is text and is either 'male' or 'female'.  Any knowledge of the history of the titanic will likely include the phrase 'women and children first'.  We should expect that it is probably a factor in survival - and we can validate it pretty quickly.

Age is self explanatory - it has a lot of missing values - we will look at mechanisms for filling in these values from the rest of the data, we expect that age will be a factor in survival - but need to explore how it affects it - expectation is probably that the elderly may have been less likely to survive - but despite being less able we expect the young may have survived both from having an adult caring for them and from the 'women and children first' policy - which is as yet unvalidated.

Lets look at our spread of values in the numeric columns using the describe method.

The PassengerID is just an incrementing counter which runs continuously from the knowns 1 through 891 and the unknowns (892 through 1309) - so the numerical analysis is pretty meaningless.

The counts are the same as above, the missing values in the ages column are clear.

In [9]:
print ("Known :")
print (titanic_known.describe())
print ("\nUnknown :") 
print (titanic_unknown.describe())

Known :
       PassengerId    Survived      Pclass  ...       SibSp       Parch        Fare
count   891.000000  891.000000  891.000000  ...  891.000000  891.000000  891.000000
mean    446.000000    0.383838    2.308642  ...    0.523008    0.381594   32.204208
std     257.353842    0.486592    0.836071  ...    1.102743    0.806057   49.693429
min       1.000000    0.000000    1.000000  ...    0.000000    0.000000    0.000000
25%     223.500000    0.000000    2.000000  ...    0.000000    0.000000    7.910400
50%     446.000000    0.000000    3.000000  ...    0.000000    0.000000   14.454200
75%     668.500000    1.000000    3.000000  ...    1.000000    0.000000   31.000000
max     891.000000    1.000000    3.000000  ...    8.000000    6.000000  512.329200

[8 rows x 7 columns]

Unknown :
       PassengerId      Pclass         Age       SibSp       Parch        Fare
count   418.000000  418.000000  332.000000  418.000000  418.000000  417.000000
mean   1100.500000    2.265550   30.272590   

We have two groups of passengers - before we do anything with the data let's just check that the two sets are reasonably comparable.  By looking a a simple correlation matrix between the numeric collumns we should be able to see if the two groups are basically similar.

We'll just run the correlation for the selected numeric columns - and we'll skip the 'Survived' column for the moment.  

Clearly the Survived column's correlation with the other data is key to what we are trying to do - so we will focus on it when we have completed our exploration of the basic data.

In [10]:
print ("Known :")
print (titanic_known[['Pclass', 'Age', 'SibSp', 'Parch', 'Fare']].corr())
print ("\nUnknown :") 
print (titanic_unknown[['Pclass', 'Age', 'SibSp', 'Parch', 'Fare']].corr())

Known :
          Pclass       Age     SibSp     Parch      Fare
Pclass  1.000000 -0.369226  0.083081  0.018443 -0.549500
Age    -0.369226  1.000000 -0.308247 -0.189119  0.096067
SibSp   0.083081 -0.308247  1.000000  0.414838  0.159651
Parch   0.018443 -0.189119  0.414838  1.000000  0.216225
Fare   -0.549500  0.096067  0.159651  0.216225  1.000000

Unknown :
          Pclass       Age     SibSp     Parch      Fare
Pclass  1.000000 -0.492143  0.001087  0.018721 -0.577147
Age    -0.492143  1.000000 -0.091587 -0.061249  0.337932
SibSp   0.001087 -0.091587  1.000000  0.306895  0.171539
Parch   0.018721 -0.061249  0.306895  1.000000  0.230046
Fare   -0.577147  0.337932  0.171539  0.230046  1.000000


This shows that our two groups are somewhat similar - but we have only looked at the numeric values.  This correlation ignores all the missing numeric values - and we know that a lot of ages are missing - a simple way to extend this analysis would be to encode gender and embarkation as integers and this would give us some additional information.

However we will be doing this anyway at a later stage so we can look at the revised correlation then.  The purpose of this exercise is just familiarity with the data.



Let's go and have a look at these age values - there are a lot missing - let's look at how best to supply them.

First how many are missing  in each of our datasets?

In [19]:

print("Known")
print(sum(pd.isnull(titanic_known['Age'])))
print( str(round(sum(pd.isnull(titanic_known['Age']))/(len(titanic_known["PassengerId"]))*100,2)) + " %")

print("Unknown")
print(sum(pd.isnull(titanic_unknown['Age'])))
print( str(round(sum(pd.isnull(titanic_unknown['Age']))/(len(titanic_unknown["PassengerId"]))*100,2)) + " %")

Known
177
19.87 %
Unknown
86
20.57 %


OK - so pretty much the same percentage missing in each case - which is good as we want the knowns and unknowns to be similar othewise we have a bigger problem to solve.

And let's look at the correlation between Age and the other numeric values to help understand the data:

And we're just going to pullout the Age row from the correlation dataframe - so it comes out as a series:

In [29]:
print ("Known :")
print (titanic_known[['Age','Pclass', 'SibSp', 'Parch', 'Fare','Survived']].corr().loc['Age',:])

Known :
Age         1.000000
Pclass     -0.369226
SibSp      -0.308247
Parch      -0.189119
Fare        0.096067
Survived   -0.077221
Name: Age, dtype: float64




We have a choice of approaches here - we can fill in the knowns with a variety of methods and see which one gives us the better result - but the differences between the aproaches to filling the data will probably be hidden by other effects. Alternatively we can treat both datasets together and attempt to fill in the values in both at the same time - which precludes using survival as a factor in filling in age - which is a somewhat circular approach anyway.

In terms of approches to the actual value there are single value approaches - populate all the missing values with a single value - and typically these will use either the mean (for regular distributed data) or the median (for skewed data) - or some other justifiable value based on the data.  And there are multiple value approaches which in turn divide down into subset versions of the single value approach - for example in this case take the mean of ages for each class and apply it to the missing values for that class (note the correlation between age and class above as a justification for this approach) - similarly for gender, for example, or get sophisticated and divide into gender and class and fill each with the relevant mean, which we can't currenlty justify until we know whether there is a supporting correlation.  

And finally there are existing algorithms for filling in missing data using the other data for each passenger which is really just a sophisticated version of the divide and fill just mentioned.  Of these MICE (Multiple (or Multivariate) Imputation by Chained Equations) is the most popular sophisticated date filling algorithm in use at the moment and there are pythons libraries to support it.

For the moment we will use a dumb approach across the entire dataset - and simply apply a rounded mean age.

The reason to not use MICE at this point is that we will tease out some more data - specifically we can pull out the titles and use these as additional input - for male passengers we know that 'Master' vs 'Mr.' will correlate with age, without even having to look at the data and to a lesser extent 'Miss' vs 'Mrs' will do so for females - the underlying argument here is that to make good use of a sophisticated tool you should invest in good data prep and make it pay. 

So single value for knowns and unknowns or one for each - much of this exercise is based on an assumption that the unknowns are a good selection from the data, which we have partially validated with the correlation exercise above so the mean of all the supplied ages used to fill all the missing ones is the initial cheap and easy approach.

And we'll come back around to this later to do a better job.

Just one additional point to consder here - our currrent problem is to find ways to populate the survived column in the unknowns based on the available data and using the known survived to help us build and test the aproaches.  We could turn this problem around and try to find the ages for all passengers in a data set where there is a boollean collum of survived which happens to have a bunch of missing values.  So data wrangling to prepare your data is in some case the same class of a problem as the actual analytics exercise itself.  And there is very little point in throwing a sophisticated analytics method at data that you have already polluted by using low tech preparation tools on.